app/src/main/java/ch/epfl/qedit/view/treasurehunt/QuestionLocatorActivity.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    private void askPermissions() {
+        permManager.requestPermissions(this, this, REQUESTED_PERMISSIONS);
+    }
+
+    @Override
+    public void onPermissionResult(String[] permissions, boolean[] granted) {
+        for (int i = 0; i < granted.length; i++) {
+            if (!granted[i]) {
+                // If the permissions were not granted, we just show an error UI to the user
+                // If it is the first time the user denies, we show the button, otherwise we
+                // just give up and show nothing.
+                if (permManager.shouldAskAgain(this, permissions[i])) setPermissionUI();
+                else button.setVisibility(View.GONE);
+
+                return;
+            }
+        }
+
+        // The user finally accepted, so we can hide the button and subscribe to the location
+        locService.subscribe(this, LOCATION_INTERVAL);
+        button.setVisibility(View.GONE);
+    }
+
+    // Sets the text of the UI to inform the user that his position is unknown
+    private void setUnknownUI() {
+        text1.setText(getString(R.string.question_locator_error));
+        text2.setText("");
+    }
+
+    // Sets the button of the UI so that the user can allow the location
+    private void setPermissionUI() {
+        button.setOnClickListener(v -> askPermissions());
+        button.setText(R.string.enable_location);
+        button.setVisibility(View.VISIBLE);
+    }
+
+    // This method shows the final UI that appears when the user found the question
+    private void setFinishUI() {
+        text1.setText(getString(R.string.question_locator_found));
+        text2.setText("");
+        button.setOnClickListener(
+                v -> Toast.makeText(this, "Not yet implemented", Toast.LENGTH_SHORT).show());
+        button.setText(getString(R.string.question_locator_answer));
+        button.setVisibility(View.VISIBLE);
+    }
+
+    // This method is used to prevent the modulo from returning negative numbers
+    private static float mod(float a, float b) {
+        float m = a % b;
+        return m < 0 ? m + b : m;
+    }
+
+    @Override
+    // This method is called when the location service sends location information
+    public void onLocationChanged(Location location) {
+        float distance = location.distanceTo(questionLoc);
+        float targetBearing = location.bearingTo(questionLoc);
+        float currentBearing = location.getBearing();
+
+        updateUI(distance, mod(targetBearing, 360), mod(targetBearing - currentBearing, 360));
+    }
+
+    // This function is used to ensure that the modulo always return a positive number
+    // This method returns a user friendly distance indication from a distance measured in meters
+    private static String formatDistance(float distance) {
+        // Note: \u00A0 is a non-breaking space
+        if (distance < 1000) return String.format("%.0f\u00A0m", distance);
```

Comment:
Maybe you could use a string constant for the non-breaking space

Reply:
Mhh yeah, but it is the only place I use it so I will keep it here for now.